README.md
pint
====

Probabilistic dependency analysis for functional genomics.
In particular, tools for integrative screening
of genome-wide RNA and DNA profiling data for cancer gene discovery 
and functional analysis of chromosomal aberrations Lahti et al. MLSP
2009.

The package implements probabilistic models for integrative analysis
of mRNA expression levels with DNA copy number (aCGH) measurements to
discover functionally active chromosomal alterations. The algorithms
can be used to discover functionally altered chromosomal regions and
to visualize the affected genes and samples. The algorithms can be
applied also to other types of biomedical data, including epigenetic
modifications, SNPs, alternative splicing and transcription factor
binding, or in other application fields. By investigating dependencies
between different functional layers of the genome it is possible to
discover mechanisms and interactions that are not seen in the
individual measurement sources. For instance, integration of gene
expression and DNA copy number can reveal cancer-associated
chromosomal regions and associated genes with potential diagnostic,
prognostic and clinical impact.

The methods are based on latent variable models including
probabilistic canonical correlation analysis and related extensions,
implemented in the [dmt
package](http://cran.fhcrc.org/web/packages/dmt/index.html).
Probabilistic formulation deals rigorously with uncertainty associated
with small sample sizes common in biomedical studies and provides
tools to guide dependency modeling through Bayesian priors.

This is the development version of the package. For a stable release
version, see
[Bioconductor](http://bioconductor.org/packages/release/bioc/html/pint.html).
For further details and examples, see the package
[vignette](vignettes/depsearch.Rnw).




inst/NEWS
CHANGES IN VERSION 1.37.2 (2020-01-12)
  o Bridged with git and resolved the most critical parts to maintain package functionality

CHANGES IN VERSION 1.7.03 (2012-03-22)

  o fixed missing documentations

CHANGES IN VERSION 1.7.01 (2012-01-16)

  o added remove.duplicates option to pint.match

CHANGES IN VERSION 1.5.35 (2011-10-13)

  o Fixed bugs with GenomeModels accessors

CHANGES IN VERSION 1.5.34 (2011-08-15)

  o changed the default match.probes to TRUE in screen.cgh.mrna; now
    also checks with match.probes = FALSE that the probes are matched

CHANGES IN VERSION 1.5.33 (2011-08-02)

  o now artificially adding arm information if it is missing, to avoid
    halts

CHANGES IN VERSION 1.5.30 (2011-07-15)

  o screen.cgh.mrna option segmented changed into match.probes

CHANGES IN VERSION 1.5.29 (2011-05-09)

BUG FIXES

  o Fixed a bug in plot.ChromosomeModels

CHANGES IN VERSION 1.5.28 (2011-05-09)

BUG FIXES

  o fixed a bug which prevented screeningn while missing values were present

CHANGES IN VERSION 1.5.06 (2011-04-16)

  o added interpretation tools in report.R

CHANGES IN VERSION 1.4.02 (2011-04-16)

  o still correcting critical bug in fixed.window and associated
    functions: now accepting data with no chromosomal arm information

CHANGES IN VERSION 1.4.01 (2011-04-14)

  o corrected critical bug in fixed.window and associated functions

CHANGES IN VERSION 1.3.85 (2011-04-11)

  o Added a method to get the latent variable Z for particular model

CHANGES IN VERSION 1.3.84 (2011-04-10)

  o vignette updated

CHANGES IN VERSION 1.3.83 (2011-04-10)

BUG FIXES

  o Multiple bugs with GenomeModels methods fixed

CHANGES IN VERSION 1.3.82 (2011-04-10)

BUG FIXES

  o plot.ChromosomeArmModels now prints by default the arm in the 
    plot title if only models from one arm is plotted

CHANGES IN VERSION 1.3.80 (2011-03-27)

SIGNIFICANT USER-VISIBLE CHANGES

  o Dependency modeling has been splitted to dmt package.

  o ChromosomeArmModels class is removed from the package.

CHANGES IN VERSION 1.3.71 (2011-03-10)

  o The sign of the effects in z.effects and W.effects now corresponds with the data

CHANGES IN VERSION 1.3.70 (2011-03-04)

BUG FIXES

  o fixed bug which caused an error when using findModel to return a model which is 
    in the end of a chromosome


CHANGES IN VERSION 1.3.69 (2011-03-02)

  o added as.data.frame method

BUG FIXES

  o plot method for GenomeModels now works with showTop argument

CHANGES IN VERSION 1.3.68 (2011-03-02)

  o corrected bibtex bug in doc

CHANGES IN VERSION 1.3.67 (2011-02-01)

  o if row or column names for the data are missing, they are named by
    indices and warning is given (pint.data function)

CHANGES IN VERSION 1.3.66 (2011-01-29)

BUG FIXES

  o pint.data: dimnames now provided for the output data and info fields

CHANGES IN VERSION 1.3.63 (2011-01-13)

BUG FIXES

  o as.data.frame fixed

CHANGES IN VERSION 1.3.63 (2010-01-11)

  o Added many new development features that help to handle segmented
    data (options segmented and nonmatched, and related scripts). Will
    be documented properly in later versions.

CHANGES IN VERSION 1.3.35 (2010-12-30)

BUG FIXES 

  o in pint.match added rownames check between Y$info and
    Y$data, previously only number of rows was checked for Y data. For
    X data things were ok: both row number and row names were checked

  o in pint.data call within pint.match, the X data set was
    erroneously used for both X and Y calls. Now calling Y for Y and X
    for X. Also polishing replace.inf part. Now also naming the info field
    rows correctly.
    
NEW FEATURES
   
  o checking info fields in pint.data for synonymous terms: replacing
    "chromosome" with "chr" and "stop" with "end". Requires less
    tweaking from the end user.

  o if chromosome names X or Y used, replace these with 23 and
    24. This makes code compatible with any organism.

MISC
 
  o changed the internal function name 'impute' to 'imputation'

  o moved some functionality from pint.match to pint.data

CHANGES IN VERSION 1.3.33 (2010-12-09)

BUG FIXES

 o corrected bug in impute.R function which produced NAs

 o added replace.inf in pint.data; now this preprocessing function
   (and consequently pint.match) replaces infinite values with large
   values by default

CHANGES IN VERSION 1.3.32 (2010-12-09)

NEW FEATURES

 o imputation option added (default TRUE) to pint.match and improved
   in pint.data
 

CHANGES IN VERSION 1.3.3 (2010-10-21)

NEW FEATURES

  o GenomeModels and ChromosomeModels classes have getModelNumbers
  method
  o topGenes and topModels methods have default values
  o screen.cgh.mrna checks whether data has been segmented and 
  warns the user

CHANGES IN VERSION 1.3.2 (2010-10-20)

NEW FEATURES

  o Default value for windowSize argument in screening function 
    changed

CHANGES IN VERSION 1.3.1 (2010-10-19)

NEW FEATURES

  o Segmented gene copy number data can be screened with pSimCCA
    method, although it is not recommended.
  o Output type of screening functions can changed to a data frame.
  o All dependency model outputs from screening functions can now
    converted to a data frame with as.data.frame method.
  o ChromosomeArmModels can be accessed with [['p']] and [['q']]
    from ChromosomeModels.

BUG FIXES

  o z.effects gives a proper error message when the window size is
    too large.
  o pCCA method gives a proper error message when the data matrix is not
    invertible.

CHANGES IN VERSION 1.1.17 (2010-10-15)

HELP & Rd FILE CHANGES

  o sample and genes are displayed in correct order in help pages
    for screening functions

CHANGES IN VERSION 1.1.16 (2010-08-04)

BUG FIXES

  o screening functions can handle chromosomes with fewer samples
  than window size
  o isEmpty method is now exported
  o topModels method for ChromosomeModels class gives no longer
  an error when either arm has no dependency models

CHANGES IN VERSION 1.1.15 (2010-06-16)


R/screen.cgh.mrna.R
#' @title Screen Associations
#' @aliases screen.cgh.mrna, screen.cgh.mir
#' @description Fits dependency models to chromosomal arm, chromosome or the whole genome.
#' @param X Data sets 1. It is recommended to place gene/mirna expression data in X and copy number data in Y. Each is a list with the following items: data: in a matrix form. Genes are in rows and samples in columns. e.g. gene copy number; info: Data frame which contains following information aboout genes in data matrix; chr: Number indicating the chrosome for the gene: (1 to 24). Characters 'X' or 'Y' can be used also.; arm: Character indicating the chromosomal arm for the gene ('p' or 'q'); loc: Location of the gene in base pairs; pint.data: can be used to create data sets in this format.
#' @param Y The second data item.
#' @param windowSize Determine the window size. This specifies the number of nearest genes to be included in the chromosomal window of the model, and therefore the scale of the investigated chromosomal region. If not specified, using the default ratio of 1/3 between features and samples or \code{15} if the ratio would be greater than 15
#' @param chromosome Specify the chromosome for model fitting. If missing, whole genome is screened.
#' @param arm Specify chromosomal arm for model fitting. If missing, both arms are modeled.
#' @param method Dependency screening can utilize any of the functions from the package dmt (CRAN). Particular options include 'pSimCCA' probabilistic similarity constrained canonical correlation analysis \cite{Lahti et al. 2009}. This is the default method.; 'pCCA' probabilistic canonical correlation analysis Bach & Jordan 2005; 'pPCA' probabilistic principal component analysis Tipping & Bishop 1999; 'pFA': probabilistic factor analysis \cite{Rubin & Thayer 1982}; 'TPriorpSimCCA' probabilistic similarity constrained canonical correlation analysis with possibility to tune T prior (Lahti et al. 2009)
#' @param param
#' List of parameters for the dependency model.
#'       sigmas: Variance parameter for the matrix normal prior
#' 	distribution of the transformation matrix T. This describes the
#' 	deviation of T from H
#'       H: Mean parameter for the matrix normal prior distribution
#' 	prior of transformation matrix T
#'       zDimension: Dimensionality of the latent variable
#'       mySeed: Random seed.
#'       covLimit: Convergence limit. Default depends on the selected
#' 	method: 1e-3 for pSimCCA with full marginal covariances and 1e-6	
#' for pSimCCA in other cases.
#'    max.dist: Maximum allowed distance between probes. Used in
#'     automated matching of the probes between the two data sets based on
#'     chromosomal location information.
#'    outputType: Specifies the output type of the function. possible values are \code{"models"} and \code{"data.frame"}
#' @param useSegmentedData: Logical. Determines the useage of the method for segmented data
#' @param match.probes To be used with segmented data, or nonmatched probes in general. Using nonmatched features (probes) between the data sets. Development feature, to be documented later.
#' @param regularized Regularization by nonnegativity constraints on the projections. Development feature, to be documented later.
#' @details Function \code{screen.cgh.mrna} assumes that data is already
#'  paired. This can be done with \code{\link{pint.match}}. It takes sliding
#' gene windows with \code{\link{fixed.window}} and fits dependency models
#' to each window with \code{\link{fit.dependency.model}} function. If the
#' window exceeds start or end location (last probe) in the chromosome in
#' the \code{\link{fixed.window}} function, the last window containing the
#' given probe and not exceeding the chromosomal boundaries is used. In
#' practice, this means that dependency score for the last n/2 probes in
#' each end of the chromosome (arm) will be calculated with an identical
#' window, which gives identical scores for these end position probes. This
#' is necessary since the window size has to be fixed to allow direct
#' comparability of the dependency scores across chromosomal windows.
#' Function \code{screen.cgh.mir} calculates dependencies
#' around a chromosomal window in each sample in \code{X}; only one sample
#' from \code{X} will be used. Data sets do not have to be of the same size
#' and\code{X} can be considerably smaller. This is used with e.g.  miRNA
#' data. If method name is specified, this overrides the corresponding model
#' parameters, corresponding to the modeling assumptions of the specified
#' model. Otherwise method for dependency models is determined by
#' parameters. Dependency scores are plotted with \link{dependency score plotting}.
#' @return The type of the return value is defined by the the function argument \code{outputType}. With the argument \code{outputType = "models"}, the return value depends on the other arguments; returns a \linkS4class{ChromosomeModels} which contains all the models for dependencies in chromosome or a \linkS4class{GenomeModels} which contains all the models for dependencies in genome. With the argument \code{outputType = "data.frame"}, the function returns a data frame with eachs row representing a dependency model for one gene. The columns are: geneName, dependencyScore, chr, arm, loc
#' @references
#' Dependency Detection with Similarity Constraints, Lahti et al., 2009
#' Proc. MLSP'09 IEEE International Workshop on Machine Learning for Signal
#' Processing, See
#' \url{http://www.cis.hut.fi/lmlahti/publications/mlsp09_preprint.pdf}
#'
#' A Probabilistic Interpretation of Canonical Correlation Analysis, Bach
#' Francis R. and Jordan Michael I. 2005 Technical Report 688. Department
#' of Statistics, University of California, Berkley.
#' \url{http://www.di.ens.fr/~fbach/probacca.pdf}
#'
#' Probabilistic Principal Component Analysis, Tipping Michael E. and
#' Bishop Christopher M. 1999. \emph{Journal of the Royal Statistical
#' Society}, Series B, \bold{61}, Part 3, pp. 611--622.
#' \url{http://research.microsoft.com/en-us/um/people/cmbishop/downloads/Bishop-PPCA-JRSS.pdf}
#' EM Algorithms for ML Factoral Analysis, Rubin D. and Thayer
#' D. 1982. \emph{Psychometrika}, \bold{vol. 47}, no. 1.
#' @author Leo Lahti \email{leo.lahti@iki.fi}
#' @seealso
#' To fit a dependency model: \code{\link{fit.dependency.model}}.
#' \linkS4class{ChromosomeModels} holds dependency models for chromosome, 
#' \linkS4class{GenomeModels} holds dependency models for genome. For
#' plotting, see: 
#' \link{dependency score plotting}
#'
#' @export
#' @examples
#' data(chromosome17)
screen.cgh.mrna <- function(X, Y, windowSize = NULL,
                            chromosome,
                            arm,
                            method = "pSimCCA",
                            params = list(),
                            max.dist = 1e7, 
                            outputType = "models",
                            useSegmentedData = TRUE,
                            match.probes = TRUE,
                            regularized = FALSE)
{


  # FIXME: quick hack - later modify genomeModels class
  if (is.null(X$info$arm) && is.null(Y$info$arm)) {
    warning("Arm information missing, artificially adding p arm for all probes.")
    X$info$arm <- rep("p", nrow(X$info))
    Y$info$arm <- rep("p", nrow(Y$info))
  }

  if (is.null(X$info$arm) && !is.null(Y$info$arm) && nrow(X$info) == nrow(Y$info)) {
    warning("Arm information missing from X data, borrowing the arm info from Y data.")
    X$info$arm <- Y$info$arm
  }

  if (!is.null(X$info$arm) && is.null(Y$info$arm) && nrow(X$info) == nrow(Y$info)) {
    warning("Arm information missing from Y data, borrowing the arm info from X data.")
    Y$info$arm <- X$info$arm
  }    

  if (is.null(windowSize)) {
    windowSize <- min(floor(ncol(X$data)/3),15)
    cat(paste("Chromosomal window (windowSize) not specified. Using
default ratio of 1/3 between features and samples (with max window
size 15 probes). Using window size ", windowSize,"\n"))
  }
      
  #FIXME: move all preprocessing stuff to dedicated preprocessing functions
  # pint.data and pint.match
  
  # Check ordering of samples
  if (any(colnames(X$data) != colnames(Y$data))) {
    warning("Samples not in the same order in the two data sets. Using samples that are found in both data sets and reordering the samples.\n")

    commons <- intersect(colnames(X$data), colnames(Y$data))
    if (length(commons) > 1) {
      X$data <- X$data[, commons]
      Y$data <- Y$data[, commons]
    } else {
      stop("Not enough common samples found. Check that the corresponding samples in the two data sets have identical names.\n")
    }
  }

  ## Checks that segmented data is not used when not implicitely
  ## indicated by argument
  ## FIXME: this is independent of 'segmented' option, join these later
  if (!useSegmentedData){
    if (test.segmented(X$data) || test.segmented(Y$data)){
      warning("Segmented data found while method for non-segmented data is selected.\n", immediate. = TRUE)    
    }
  }

  if (match.probes) {   # FIXME: change name to match.probes or something 
    # Match probes
    tmp <- pint.match(X, Y, max.dist, useSegmentedData = useSegmentedData)
    X <- tmp$X
    Y <- tmp$Y
    match.probes <- FALSE # now the probes are matched.
  } else {
    # If user claims that no matching is needed
    # this will implicate that matching has already been
    # performed. Verify: the number of probes should be
    # identical in ge and cn data sets
    if (!nrow(X$data) == nrow(Y$data)) {
      stop("If match.probes == FALSE then the number of probes in ge and cn data sets (X$data, Y$data) need to match!")      
    }  
  }
  
  # Remove probes where observations are not available in either data set
  # TODO

  ############################################################################

  # Set priors

  # Currently imlement only pSimCCA for the screening
  if (method == "pSimCCA") {
        # similarity prior: Wx = Wy identically
	priors <- list(sigma.w = 0) 
  } else if (method == "nonmatched") {
    # no similarity prior for Wx ~ Wy, completely uncoupled 
    priors <- list(sigma.w = Inf)
  } else {
    message("Method not among pSimCCA, nonmatched. Using empty prior.")
    priors <- NULL
  }
	      
  if (regularized) {priors$W <- 1e-3} # W>=0 prior
		  
  ###########################################################################

  if (method == "pSimCCA") {
    if (is.null(params$H)) {
      params$H <- diag(1, windowSize, windowSize)
    }
  } else if (method == "pPCA" || method == "pCCA" || method == "pFA") {
    params$H <- NA
  } else {
    if (is.null(params$H))
      params$H <- diag(1, windowSize, windowSize)
  }

  if (is.null(params$sigmas))
    params$sigmas <- 0
  if (method == "pPCA") {
    params$marginalCovariances <- "identical isotropic"
  } else if (method == "pFA") {
    params$marginalCovariances <- "diagonal"
  } else if (method == "pCCA") {
    if (is.null(params$marginalCovariances)) {
      params$marginalCovariances <- "full"
    }
  } else {
    if (is.null(params$marginalCovariances)) {
      if (params$sigmas == 0) {
        params$marginalCovariances <- "full"
      } else {
        params$marginalCovariances <- "isotropic"
      }
    }
  }
  if (is.null(params$zDimension))
    params$zDimension <- 1
  if (!is.null(params$H) && any(is.na(params$H))) {
    if (params$marginalCovariances == "full")
      method = "pCCA"
    if (params$marginalCovariances == "isotropic")
      method = "pCCA"
    if (params$marginalCovariances == "diagonal")
      method = "pFA"
    if (params$marginalCovariances == "identical isotropic")
      method = "pPCA"
  } else {
    method = "pSimCCA"
  }

  # Convert X and Y chromosomes to numbers
  if (!missing(chromosome)){
    if (chromosome == 23) chromosome <- "X"
    if (chromosome == 24) chromosome <- "Y"
  }

  # give warning if arm param given and no arm data is available
  if (!missing(arm) && any(X$info$arm == "")){
    warning("No arm information in the data. Calculating thw whole chromosome")
    arm <- NULL
  }

  if (missing(chromosome)) {
    models <- calculate.genome(X, Y, windowSize, method, params, match.probes = match.probes, priors = priors, regularized = regularized)
  } else if (missing(arm) || is.null(arm)) {
    models <- calculate.chr(X, Y, windowSize, chromosome, method, params, match.probes = match.probes, priors = priors, regularized = regularized)
  } else {
    models <- calculate.arm(X, Y, windowSize, chromosome, arm, method, params, match.probes = match.probes, priors = priors, regularized = regularized)
  }

  # TODO: move this when segmented method is fully implemented (option 'match.probes')
  models@params <- c(models@params, segmentedData = useSegmentedData)

  if(outputType == "data.frame"){
    message("Convert model to data.frame")
    return(as.data.frame(models))
  }
  else {
    return(models)
  }
}

